core/lib/Cardano/Address/Style/Shared.hs

```diff
@@ -0,0 +1,323 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE BinaryLiterals #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveFunctor #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+
+{-# OPTIONS_HADDOCK prune #-}
+
+-- |
+-- Copyright: © 2018-2021 IOHK
+-- License: Apache-2.0
+
+module Cardano.Address.Style.Shared
+    ( -- $overview
+
+      -- * Shared
+      Shared
+    , getKey
+    , liftXPrv
+    , liftXPub
+
+      -- * Key Derivation
+      -- $keyDerivation
+    , genMasterKeyFromXPrv
+    , genMasterKeyFromMnemonic
+    , deriveAccountPrivateKey
+    , deriveAddressPrivateKey
+    , deriveAddressPublicKey
+    , deriveDelegationPrivateKey
+    , deriveDelegationPublicKey
+    , hashKey
+
+    ) where
+
+import Prelude
+
+import Cardano.Address.Derivation
+    ( Depth (..)
+    , DerivationScheme (..)
+    , DerivationType (..)
+    , Index (..)
+    , XPrv
+    , XPub
+    , deriveXPrv
+    , deriveXPub
+    , generateNew
+    , hashCredential
+    , indexFromWord32
+    , xpubPublicKey
+    )
+import Cardano.Address.Script
+    ( KeyHash (..), KeyType )
+import Cardano.Address.Style.Shelley
+    ( Role (..), coinTypeIndex, minSeedLengthBytes, roleToIndex )
+import Cardano.Mnemonic
+    ( SomeMnemonic, someMnemonicToBytes )
+import Control.DeepSeq
+    ( NFData )
+import Control.Exception.Base
+    ( assert )
+import Data.ByteArray
+    ( ScrubbedBytes )
+import Data.Maybe
+    ( fromMaybe )
+import Data.Word
+    ( Word32 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Address.Derivation as Internal
+import qualified Data.ByteArray as BA
+
+-- $overview
+--
+-- This module provides an implementation of:
+--
+-- - 'Cardano.Address.Derivation.GenMasterKey': for generating Shared master keys from mnemonic sentences
+-- - 'Cardano.Address.Derivation.HardDerivation': for hierarchical hard derivation of parent to child keys
+-- - 'Cardano.Address.Derivation.SoftDerivation': for hierarchical soft derivation of parent to child keys
+--
+-- - 'paymentAddress': for constructing payment addresses from a address public key or a script
+-- - 'delegationAddress': for constructing delegation addresses from payment credential (public key or script) and stake credential (public key or script)
+-- - 'pointerAddress': for constructing delegation addresses from payment credential (public key or script) and chain pointer
+-- - 'stakeAddress': for constructing reward accounts from stake credential (public key or script)
+
+-- | A cryptographic key for sequential-scheme address derivation, with
```

Comment:
It would be informative to describe this in terms of the shelley scheme. E.g. that it's mostly like Shelley, _except_ ....